### PR TITLE
ASIM Cisco ISE AuditEvent Schema Parser

### PR DESCRIPTION
**Pull request by spsocprime**
Required items, please complete
   
   Change(s):
   - Added Cisco ISE parser for ASIM AuditEvent Schema
   - Added test results
   - Added Sample Data file  to 'Sample Data/ASIM' folder

   Reason for Change(s):
   - Initial Commit

   Testing Completed:
   - Yes

### CODE REVIEW

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+              , TargetAppName = "ISE"
+              , TargetAppType = "Service"
+      | parse-kv SyslogMessage as (NetworkDeviceName:string, ['User-Name']:string, UserName:string, User:string, ['Remote-Address']:string, ['Device IP Address']:string) with (pair_delimiter=',', kv_delimiter='=')
+      | project-rename
+                    SrcIpAddr=['Remote-Address']
+                    , TargetIpAddr =['Device IP Address']
+    | extend DvcHostname = coalesce(NetworkDeviceName, Computer, HostName)
+    | extend ActorUsername = coalesce(['User-Name'], UserName, User)
+    | extend ActorUsernameType = ASIM_GetUsernameType(ActorUsername)      
+    | project-away NetworkDeviceName, ['User-Name'], UserName, User
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml:63]: can we do this together at line 74?

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml [hunk 2]
```diff
@@ -0,0 +1,137 @@
+Parser:
+  Title: Audit Event ASIM parser for Cisco ISE
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml:2]: Audit Event ASIM filtering parser for Cisco ISE

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  Version: '1.0'
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml:3]: All comments applicable for VIM parser too as provided in ASim parser.

File: Parsers/ASimAuditEvent/Tests/Cisco_ISE_AuditEvent_SchemaTest.csv [hunk 4]
```diff
@@ -0,0 +1,90 @@
+﻿Result
+"(1) Warning: Missing recommended field [DvcAction]"
+"(1) Warning: Missing recommended field [DvcDomain]"
+"(1) Warning: Missing recommended field [EventResultDetails]"
+"(1) Warning: Missing recommended field [EventUid]"
+"(1) Warning: Missing recommended field [NewValue]"
+"(1) Warning: Missing recommended field [ObjectId]"
+"(1) Warning: Missing recommended field [TargetHostname]"
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Tests/Cisco_ISE_AuditEvent_SchemaTest.csv:8]: are we not getting these recommended fields from ISE?

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+          username contains "@" , "UPN"
+          , username contains "\\", "Windows"
+          , isempty(username), ""
+          , "Simple"
+      )
+    };
+    let CiscoISEAuditParser=(disabled:bool=false){
+      Syslog
+      | where not(disabled)
+      | where ProcessName has_any ("CISE", "CSCO")
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml:31]: This is according to the official documentation:
https://www.cisco.com/c/en/us/td/docs/security/ise/syslog/Cisco_ISE_Syslogs/m_IntrotoSyslogs.html

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      | extend EventOriginalType = extract(@".*\d{10}\s(\d{4,5})\s.*", 1, SyslogMessage)
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml]: fixed

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
..."60060","60061","60062","60063","60064","60068","60069","60070","60083","60084","60085","60086","60087","60088","60089","60090","60091","60092","60094","60095","60097","60098","60100","60101","60102","60103","60105","60106","60107","60111","60112","60113","60114","60118","60119","60120","60121","60122","60123","60126","60128","60129","60130","60131","60132","60133","60153","60166","60167","60172","60173","60187","60189","60193","60194","60197","60198","60199","60200","60201","60202","60207","60208","60209","60210","60213","60214","60215","60216","60451","60452","61002","61003","61004","61005","61006","61007","61008","61009","61016","61017","61018","61020","61022","61023","61024","61030","61033","61034","61051","61052","61100","61101","61102","61103","61104","61105","61106","61107","61108","61109","61110","61111","61114","61115","61116","61117","61118","61120","61121","61122","61123","61124","61125","61126","61127","61128","61129","61130","61158","61160","61161","61162","61163","61164","61165","61166","61167","61168","61169","61234","62000","62001","62002","62003","61300","61301","61302","60467","60468","62008","62009","62010","62011","62012","62013","62014","62015","62016","62017")
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml]: changed to dynamic type variable

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+      | extend 
+          EventVendor = "Cisco"
+          , EventProduct = "ISE"
+          , EventProductVersion = "3.2"
+          , EventCount = int(1)
+          , EventSchema = "AuditEvent"
+          , EventSchemaVersion = "0.1.0"
+          , EventStartTime = coalesce(EventTime, TimeGenerated)
+          , EventEndTime = coalesce(EventTime, TimeGenerated)
+          , EventOriginalSeverity = extract(@".*(NOTICE|INFO|WARN|WARNING|ERROR|FATAL|DEBUG).*", 1, SyslogMessage)
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml]: added values to lookup table

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
...arts","62009","Other","Success","Informational","Meraki connector","Meraki connector sync service stops","Meraki connector sync service stops","62010","Other","Failure","Low","Meraki connector","Meraki connector sync service failure","Meraki connector sync service failure","62011","Other","Success","Informational","Meraki connector","Meraki connector sync cycle starts","Meraki connector sync cycle starts","62012","Other","Success","Informational","Meraki connector","Meraki connector sync cycle stops","Meraki connector sync cycle stops","62013","Other","Failure","Low","Meraki connector","Meraki connector sync cycle failure","Meraki connector sync cycle failure","62014","Other","Success","Informational","Meraki connector","Meraki connector sync operation success","Meraki connector sync operation success","62015","Other","Failure","Low","Meraki connector","Meraki connector sync operation failure","Meraki connector sync operation failure","62016","Other","Success","Informational","ISE instance","Port 2484 opened for Data Connect","Port 2484 opened for Data Connect","62017","Other","Success","Informational","ISE instance","Data Connect port 2484 closed","Data Connect port 2484 closed"];
```
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml]: I didn't see this change made.

[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml]: applied formating

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+    | lookup EventFieldsLookup on EventOriginalType
+    | parse-kv SyslogMessage as (NetworkDeviceName: string, ['User-Name']: string, UserName: string, User: string, ['Remote-Address']: string, ['Device IP Address']: string) with (pair_delimiter=',', kv_delimiter='=')
+    | project-rename
+        SrcIpAddr=['Remote-Address']
+        , TargetIpAddr =['Device IP Address']
+    | where (array_length(srcipaddr_has_any_prefix) == 0 or has_any_ipv4_prefix(SrcIpAddr, srcipaddr_has_any_prefix))
+    | extend DvcHostname = coalesce(NetworkDeviceName, Computer, HostName)
+    | extend ActorUsername = coalesce(['User-Name'], UserName, User)
+    | extend ActorUsernameType = _ASIM_GetUsernameType(ActorUsername)      
+    | where (array_length(actorusername_has_any) == 0 or ActorUsername has_any (actorusername_has_any))
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml:316]: Here we filter on the `ActorUsername` which was not available during pre-filtering. Do we need to delete this filtering from the parser?

File: Parsers/ASimAuditEvent/Tests/Cisco_ISE_AuditEvent_SchemaTest.csv [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+"(2) Info: Missing optional field [DvcScope]"
+"(2) Info: Missing optional field [DvcZone]"
+"(2) Info: Missing optional field [EventOriginalResultDetails]"
+"(2) Info: Missing optional field [EventOriginalSubType]"
+"(2) Info: Missing optional field [EventOriginalUid]"
+"(2) Info: Missing optional field [EventOwner]"
+"(2) Info: Missing optional field [EventReportUrl]"
+"(2) Info: Missing optional field [EventSubType]"
+"(2) Info: Missing optional field [HttpUserAgent]"
+"(2) Info: Missing optional field [OldValue]"
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Tests/Cisco_ISE_AuditEvent_SchemaTest.csv:40]: we have not observed such data in the events

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+      | extend DvcHostname = coalesce(NetworkDeviceName, Computer, HostName)
+      | extend ActorUsername = coalesce(['User-Name'], UserName, User)
+      | extend ActorUsernameType = _ASIM_GetUsernameType(ActorUsername)      
+      | extend
+          EventStartTime = coalesce(EventTime, TimeGenerated)
+          , EventEndTime = coalesce(EventTime, TimeGenerated)
+      | extend DvcIpAddr = extract(@"(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})", 1, Computer)
+          , EventVendor = "Cisco"
+          , EventProduct = "ISE"
+          , EventProductVersion = "3.2"
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml]: we have not observed such data in the events

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    let CiscoISEAuditParser=(disabled:bool=false){
+      Syslog
+      | where not(disabled)
+      | where ProcessName has_any ("CISE", "CSCO")
+      | parse kind = regex SyslogMessage with @"\d{10}\s" EventOriginalType @"\s(NOTICE|INFO|WARN|WARNING|ERROR|FATAL|DEBUG)"
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml]: Thank for suggesting solution. Added to parser

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+        endtime = endtime,
+        srcipaddr_has_any_prefix = srcipaddr_has_any_prefix,
+        actorusername_has_any = actorusername_has_any,
+        eventtype_in = eventtype_in,
+        eventresult = eventresult,
+        operation_has_any = operation_has_any,
+        object_has_any=object_has_any,
+        newvalue_has_any=newvalue_has_any,
+        disabled=disabled
+    )
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml:359]: Please re-test filters I found issues with the following filters:
eventtype_in not working
eventresult

Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml:360]: fixed

**Comment by v-atulyadav** [comment]
Hi @spsocprime,
Thank you for these fixes. We will provide an update as soon as possible.

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 15]
```diff
@@ -0,0 +1,295 @@
+Parser:
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml:1]: can you please add entry for this parser in union parser?
https://github.com/Azure/Azure-Sentinel/blob/master/Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml
https://github.com/Azure/Azure-Sentinel/blob/master/Parsers/ASimAuditEvent/Parsers/imAuditEvent.yaml

Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml:1]: rest all looks good

Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml:1]: Added records.
P.S. I wonder if the records are correct in `imAuditEvent.yaml` unified parser (see screenshot below). Shouldn't they start with `vim`?
image

Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml:1]: I'll look into this, can you follow the existing format for now like other audit parsers?

Also, please add ',' at the end of AzureActivity parser

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml [hunk 16]
```diff
@@ -36,3 +37,4 @@ ParserQuery: |
     ASimAuditEventMicrosoftExchangeAdmin365  (BuiltInDisabled or ('ExcludeASimAuditEventMicrosoftExchangeAdmin365' in (DisabledParsers))),
     ASimAuditEventMicrosoftWindowsEvents  (BuiltInDisabled or ('ExcludeASimAuditEventMicrosoftWindowsEvents' in (DisabledParsers))),
     ASimAuditEventAzureActivity  (BuiltInDisabled or ('ExcludeASimAuditEventAzureActivity' in (DisabledParsers)))
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml:39]: missing ','

File: Parsers/ASimAuditEvent/Parsers/imAuditEvent.yaml [hunk 17]
```diff
@@ -63,3 +63,4 @@ ParserQuery: |
     ASimAuditEventMicrosoftExchangeAdmin365  (BuiltInDisabled or ('ExcludevimAuditEventMicrosoftExchangeAdmin365' in (DisabledParsers))),
     ASimAuditEventMicrosoftWindowsEvents  (BuiltInDisabled or ('ExcludevimAuditEventMicrosoftWindowsEvents' in (DisabledParsers))),
     ASimAuditEventAzureActivity  (BuiltInDisabled or ('ExcludevimAuditEventAzureActivity' in (DisabledParsers)))
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/imAuditEvent.yaml:65]: add ','

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+        SyslogMessage,
+        HostIP,
+        ProcessName,
+        ProcessID,
+        _ResourceId,
+        NetworkDeviceName,
+        ['User-Name'],
+        UserName
+    };
+    CiscoISEAuditParser(disabled)
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml:295]: CiscoISEAuditParser(disabled=disabled)

File: Parsers/ASimAuditEvent/Parsers/imAuditEvent.yaml (hunk 17 above)
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/imAuditEvent.yaml]: added missing ','

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml (hunk 18 above)
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml]: changed

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml (hunk 16 above)
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml]: added missing ','

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml (hunk 15 above)
Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoISE.yaml:1]: Changed to existing format like for other parsers

**Comment by v-atulyadav** [comment]
Hi @socprime,
Thanks for your response, we will provide you an update shortly.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml [hunk 19]
```diff
@@ -0,0 +1,360 @@
...
+References:
+- Title: ASIM Audit Event Schema
+  Link: https://aka.ms/ASimAuditEventDoc
+- Title: ASIM
+  Link: https:/aka.ms/AboutASIM
+- Title: Cisco ISE Security Events 
+  Link: https://www.cisco.com/c/en/us/td/docs/security/ise/3-2/admin_guide/b_ise_admin_3_2/b_ISE_admin_32_maintain_monitor.html#ID58
+Description: |
+  This ASIM parser supports normalizing administrative activity in the Cisco ISE events to the ASIM Audit Event schema.
+ParserName: ASimAuditEventCiscoISE
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml:19]: @spsocprime can you please change this to vimAuditEventCiscoISE?

Comment (spsocprime) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoISE.yaml]: changed

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**